### PR TITLE
Allow grouping by navigation property

### LINKED ISSUE #689: `hierarchies`: Grouping by navigation property doesn't work
Repro full stack test:
```ts
it.only("groups by navigation property", async function () {
  const imodelAccess = createIModelAccess(emptyIModel);
  const selectClauseFactory = createNodesQueryClauseFactory({ imodelAccess });
  const provider = createHierarchyProvider({
    imodelAccess,
    hierarchyDefinition: {
      defineHierarchyLevel: async ({ parentNode }) =>
        parentNode
          ? []
          : [
              {
                fullClassName: "BisCore.Subject",
                query: {
                  ecsql: `
                    SELECT ${await selectClauseFactory.createSelectClause({
                      ecClassId: { selector: "this.ECClassId" },
                      ecInstanceId: { selector: "this.ECInstanceId" },
                      nodeLabel: { selector: "this.CodeValue" },
                      grouping: {
                        byProperties: {
                          propertiesClassName: "BisCore.Subject",
                          propertyGroups: [
                            {
                              propertyClassAlias: "this",
                              propertyName: "Parent",
                            },
                          ],
                        },
                      },
                    })}
                    FROM BisCore.Subject [this]
                  `,
                },
              },
            ],
    },
  });
  await validateHierarchy({
    provider,
    expect: [
      NodeValidators.createForPropertyValueGroupingNode({
        label: "Unspecified",
        propertyName: "Parent",
        children: [
          NodeValidators.createForInstanceNode({
            label: "Physical element",
          }),
        ],
      }),
    ],
  });
});
```

The test fails with `BE_SQLITE_ERROR: Function 'json_object' can only be called with primitive arguments. Argument #4 is not primitive.`

The resulting ECSQL query:

```sql
SELECT 
    ec_ClassName(this.ECClassId) AS FullClassName,
    this.ECInstanceId AS ECInstanceId,
    this.CodeValue AS DisplayLabel,
    CAST(NULL AS BOOLEAN) AS HasChildren,
    CAST(NULL AS BOOLEAN) AS HideIfNoChildren,
    CAST(NULL AS BOOLEAN) AS HideNodeInHierarchy,
    json_object(
        'byProperties',
        json_object(
            'propertiesClassName',
            'BisCore.Subject',
            'propertyGroups',
            json_array(
                json_object(
                    'propertyName',
                    'Parent',
                    'propertyValue',
                    [this].[Parent]
                )
            )
        )
    ) AS Grouping,
    CAST(NULL AS TEXT) AS ExtendedData,
    CAST(NULL AS BOOLEAN) AS AutoExpand,
    CAST(NULL AS BOOLEAN) AS SupportsFiltering
FROM BisCore.Subject [this]
```

Expected result: The property grouping node should be created and its formatted value/label should match the navigation property target's label.

### PR DESCRIPTION
**Pull request by JonasDov**
closes #689

### CODE REVIEW

File: packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts [hunk 1]
```diff
@@ -475,21 +507,46 @@ function createLabelGroupingBaseParamsSelectors(byLabel: ECSqlSelectClauseLabelG
   return selectors;
 }
 
-function createPropertyGroupSelectors(propertyGroup: ECSqlSelectClausePropertyGroup) {
+async function createPropertyGroupSelectors(
+  propertyGroup: ECSqlSelectClausePropertyGroup,
+  fullClassName: string,
+  classLoader: (fullName: string) => Promise<EC.Class>,
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts:513]: what's the point to pass class name + a function that loads the class? maybe just load it beforehand and pass it here?

File: packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts [hunk 2]
```diff
@@ -475,21 +507,46 @@ function createLabelGroupingBaseParamsSelectors(byLabel: ECSqlSelectClauseLabelG
...
+
+  const propertyClass = await classLoader(fullClassName);
+
+  const property = await propertyClass.getProperty(propertyGroup.propertyName);
+  if (!property) {
+    throw new Error(`Property "${propertyGroup.propertyName}" not found in ECClass "${propertyClass.fullName}".`);
+  }
+
+  if (property.isNavigation()) {
+    const customAlias = "cAlias";
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts:530]: why `customAlias`? I'd change it to `targetAlias` and value to "target".

File: packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts [hunk 3]
```diff
@@ -475,21 +507,46 @@ function createLabelGroupingBaseParamsSelectors(byLabel: ECSqlSelectClauseLabelG
...
+    selectors.push({
+      key: "propertyValue",
+      selector: `(
+          SELECT ${await instanceLabelSelectClauseFactory.createSelectClause({ className: fullClassName, classAlias: customAlias })}
+          FROM ${fullClassName} AS ${customAlias}
+          WHERE [${customAlias}].[ECInstanceId] = [${propertyGroup.propertyClassAlias}].[${propertyGroup.propertyName}].[Id]
+        )`,
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts:537]: `fullClassName` refers to the class that contains the navigation property. It's not necessarily the same as the target class, e.g. `BisCore.GeometricElement3d.Category` nav prop points to `BisCore.SpatialCategory`.

**Review by grigasp — CHANGES_REQUESTED** [review]

File: packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts (hunk 1 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts]: Changed, now the class is being provided

File: packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts (hunk 3 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts]: Did not realise that was the case. Changed the implementation to extract target class using `property.relationshipClass.(target/source).abstractConstraint.fullName`. Added tests for checking forward/backward navigation properties

File: packages/hierarchies/src/test/internal/operators/grouping/PropertiesGrouping.test.ts [hunk 4]
```diff
@@ -873,6 +873,55 @@ describe("PropertiesGrouping", () => {
...
+        const propertyInfo: propertiesGrouping.PropertyGroupInfo = {
+          ecClass: stubbedClass,
+          previousPropertiesGroupingInfo: [],
+          propertyGroup: { propertyName: "PropertyName" },
+        };
+        const expectedGroupingNodeKey: GroupingNodeKey = {
+          type: "property-grouping:value",
+          propertyName: "PropertyName",
+          propertyClassName: "TestSchema.Class",
+          formattedPropertyValue: "",
```
Comment (grigasp) [inline packages/hierarchies/src/test/internal/operators/grouping/PropertiesGrouping.test.ts:908]: IMO, maybe the test should use the general case when there is a valid property value

File: packages/hierarchies/src/test/Utils.ts [hunk 5]
```diff
@@ -14,6 +14,7 @@ import {
   ProcessedInstanceHierarchyNode,
 } from "../hierarchies/HierarchyNode";
 import { HierarchyNodeKey } from "../hierarchies/HierarchyNodeKey";
+import { TypedValueSelectClauseProps } from "@itwin/presentation-shared/lib/cjs/shared/ecsql-snippets/ECSqlValueSelectorSnippets";
```
Comment (grigasp) [inline packages/hierarchies/src/test/Utils.ts:17]: should import from barrel

File: packages/hierarchies/src/test/Utils.ts [hunk 6]
```diff
@@ -214,6 +215,23 @@ export function createIModelAccessStub() {
...
+      classAlias: string;
+      className?: string;
+      selectorsConcatenator?: (selectors: TypedValueSelectClauseProps[], checkSelector?: string) => string;
+    }): Promise<string> {
+      return `(
+        SELECT
+         [${props.classAlias}].[CodeValue]
+        FROM [meta].[ECClassDef] AS [c]
+        WHERE [c].[ECInstanceId] =[${props.classAlias}].[ECClassId]
+      )`;
```
Comment (grigasp) [inline packages/hierarchies/src/test/Utils.ts:230]: IMO this should be much simpler e,g. `[${classAlias}].[LabelProperty]`.

File: apps/full-stack-tests/src/hierarchies/grouping/PropertiesGrouping.test.ts [hunk 7]
```diff
@@ -357,7 +357,77 @@ describe("Hierarchies", () => {
...
+              typeDefinitionId: type.id,
+              codeValue: "Physical element",
+            });
+            return { physicalElement };
+          });
+
+          const groupingParams: ECSqlSelectClausePropertiesGroupingParams = {
+            propertiesClassName: "BisCore.GeometricElement3d",
+            propertyGroups: [{ propertyName: "Category", propertyClassAlias: "this" }],
+          };
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply apps/full-stack-tests/src/hierarchies/grouping/PropertiesGrouping.test.ts]: Moved

File: packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts [hunk 8]
```diff
@@ -454,8 +455,7 @@ async function createGroupingSelector(
                 serializeJsonObject(
                   await createPropertyGroupSelectors(
                     propertyGroup,
-                    grouping.byProperties!.propertiesClassName,
-                    async (fullName: string) => getClass(imodelAccess, fullName),
+                    await getClass(imodelAccess, grouping.byProperties!.propertiesClassName),
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts]: Yeah, moved above

File: packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts [hunk 9]
```diff
@@ -519,21 +518,27 @@ async function createPropertyGroupSelectors(
...
+
+    const fullName = abstractConstraint.fullName;
+    const targetAlias = "target";
     selectors.push({
       key: "propertyValue",
       selector: `(
-          SELECT ${await instanceLabelSelectClauseFactory.createSelectClause({ className: fullClassName, classAlias: customAlias })}
-          FROM ${fullClassName} AS ${customAlias}
-          WHERE [${customAlias}].[ECInstanceId] = [${propertyGroup.propertyClassAlias}].[${propertyGroup.propertyName}].[Id]
+          SELECT ${await instanceLabelSelectClauseFactory.createSelectClause({ className: fullName, classAlias: targetAlias, selectorsConcatenator: createConcatenatedValueStringSelector })}
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts]: Changed to `createConcatenatedValueJsonSelector`, in `PropertiesGrouping.ts` we now set the formattedValue to: 
```ts
const formattedValue =
  currentProperty.propertyValue instanceof Array
    ? await formatConcatenatedValue({ value: currentProperty.propertyValue, valueFormatter })
    : await valueFormatter(TypedPrimitiveValue.create(currentProperty.propertyValue, primitiveType, koqName, extendedTypeName));
```

File: apps/full-stack-tests/src/hierarchies/learning-snippets/Migration.test.ts [hunk 10]
```diff
@@ -182,7 +182,10 @@ describe("Hierarchies", () => {
           });
           const imodelAccess = createIModelAccess(imodel);
           // __PUBLISH_EXTRACT_START__ Presentation.Hierarchies.Migration.InstanceNodesOfSpecificClassesDefinition
-          const selectClauseFactory = createNodesQueryClauseFactory({ imodelAccess });
+          const selectClauseFactory = createNodesQueryClauseFactory({
+            imodelAccess,
+            instanceLabelSelectClauseFactory: createBisInstanceLabelSelectClauseFactory({ classHierarchyInspector: imodelAccess }),
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply apps/full-stack-tests/src/hierarchies/learning-snippets/Migration.test.ts]: Changed

File: packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts [hunk 11]
```diff
@@ -13,8 +13,18 @@ import {
...
+  ECClassHierarchyInspector,
+  ECSchemaProvider,
+  ECSql,
+  getClass,
+  IInstanceLabelSelectClauseFactory,
+  parseFullClassName,
+  PrimitiveValue,
+} from "@itwin/presentation-shared";
 import { HierarchyNodeAutoExpandProp } from "./HierarchyNode";
+import { createConcatenatedValueStringSelector } from "@itwin/presentation-shared/lib/cjs/shared/ecsql-snippets";
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/NodeSelectQueryFactory.ts]: Removed import

File: packages/hierarchies/src/test/NodeSelectQueryFactory.test.ts [hunk 12]
```diff
@@ -14,21 +14,114 @@ import {
...
-    factory = createNodesQueryClauseFactory({ imodelAccess });
+    instanceLabelSelectClauseFactory = createInstanceLabelSelectClauseFactoryStub();
+    factory = createNodesQueryClauseFactory({ imodelAccess, instanceLabelSelectClauseFactory });
   });
   afterEach(() => {
     sinon.restore();
   });
 
   describe("createSelectClause", () => {
+    it(`throws when property grouping uses non-existing class alias`, async () => {
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/test/NodeSelectQueryFactory.test.ts]: Class name, changed

File: packages/hierarchies/src/test/NodeSelectQueryFactory.test.ts [hunk 13]
```diff
@@ -99,7 +192,105 @@ describe("createNodesQueryClauseFactory", () => {
       );
     });
 
+    it("creates valid clause with value props and navigation property", async () => {
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/test/NodeSelectQueryFactory.test.ts:195]: Removed stuff we don't need from the test

File: packages/hierarchies/src/test/NodeSelectQueryFactory.test.ts [hunk 14]
```diff
@@ -99,7 +192,105 @@ describe("createNodesQueryClauseFactory", () => {
...
+      );
+    });
+
     it("creates valid clause with selector props", async () => {
+      imodelAccess.stubEntityClass({
+        schemaName: "testSchema",
+        className: "testName",
+        is: async () => true,
+        properties: [{ name: "PropertyName", isNavigation: () => false } as EC.NavigationProperty],
+      });
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/test/NodeSelectQueryFactory.test.ts]: Changed to:
```ts
properties: [{ name: "PropertyName", isNavigation: () => false } as EC.Property],
```

File: packages/hierarchies/src/test/internal/operators/grouping/PropertiesGrouping.test.ts (hunk 4 above)
Comment (JonasDov) [reply packages/hierarchies/src/test/internal/operators/grouping/PropertiesGrouping.test.ts]: Changed the propertyValue to a `ConcatenatedValue`

File: packages/hierarchies/src/test/Utils.ts (hunk 5 above)
Comment (JonasDov) [reply packages/hierarchies/src/test/Utils.ts]: Removed import

File: packages/hierarchies/src/test/Utils.ts (hunk 6 above)
Comment (JonasDov) [reply packages/hierarchies/src/test/Utils.ts]: Simplified to your suggestion

**Review by grigasp — APPROVED** [review]

**Pull request merged**